src/finch/finch_notation/nodes.py

```diff
@@ -0,0 +1,396 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Any
+
+from ..algebra import element_type, query_property, return_type
+from ..symbolic import Term, TermTree
+
+
+# Base class for all Finch Notation nodes
+class NotationNode(Term):
+    pass
+
+
+class NotationTree(NotationNode, TermTree):
+    @classmethod
+    def make_term(cls, head, *children):
+        return head.from_children(*children)
+
+    @classmethod
+    def from_children(cls, *children):
+        return cls(*children)
+
+
+class NotationExpression:
+    """
+    Notation AST expression base class.
+    """
+
+    def get_type(self) -> Any:
+        """
+        Get the type of the expression.
+        """
+
+
+@dataclass(eq=True, frozen=True)
+class Literal(NotationNode, NotationExpression):
+    """
+    Notation AST expression for the literal value `val`.
+    """
+
+    val: Any
+
+    def get_type(self):
+        return type(self.val)
+
+
+@dataclass(eq=True, frozen=True)
+class Value(NotationNode, NotationExpression):
+    """
+    Notation AST expression for host code `val` expected to evaluate to a value of
+    type `type_`.
+    """
+
+    val: Any
+    type_: Any
+
+    def get_type(self):
+        return self.type_
+
+
+@dataclass(eq=True, frozen=True)
+class Variable(NotationNode, NotationExpression):
+    """
+    Notation AST expression for a variable named `name`.
+    """
+
+    name: str
+    type_: Any = None
+
+    def get_type(self):
+        return self.type_
+
+
+@dataclass(eq=True, frozen=True)
+class Call(NotationTree, NotationExpression):
+    """
+    Notation AST expression for the result of calling the function `op` on
+    `args...`.
+    """
+
+    op: Literal
+    args: tuple[NotationNode, ...]
+
+    def get_type(self):
+        arg_types = [a.get_type() for a in self.args]
+        return return_type(self.op.val, *arg_types)
+
+    @classmethod
+    def from_children(cls, op, *args):
+        return cls(op, args)
+
+    def children(self):
+        return [self.op, *self.args]
+
+
+@dataclass(eq=True, frozen=True)
+class Access(NotationTree, NotationExpression):
+    """
+    Notation AST expression representing the value of tensor `tns` at the indices
+    `idx...`.
+    """
+
+    tns: NotationNode
+    mode: NotationNode
+    idxs: tuple[NotationNode, ...]
+
+    def get_type(self):
+        # Placeholder: in a real system, would use tns/type system
+        return element_type(self.tns.get_type())
+
+    @classmethod
+    def from_children(cls, tns, mode, *idxs):
+        return cls(tns, mode, idxs)
+
+    def children(self):
+        return [self.tns, self.mode, *self.idxs]
+
+
+@dataclass(eq=True, frozen=True)
+class Read(NotationNode):
+    def children(self):
+        return []
+
+
+@dataclass(eq=True, frozen=True)
+class Update(NotationTree):
+    op: NotationNode
+
+    def children(self):
+        return [self.op]
+
+
+@dataclass(eq=True, frozen=True)
+class Increment(NotationTree):
```

Comment:
Is there a better name than `Increment`? I think that `lhs = lhs - rhs` also counts as an increment now.